BIRD Benchmark (Text-to-SQL) (#1132)
NVIDIA/NeMo-Skills: docs/evaluation/code.md

--benchmarks=livecodebench:3
```

### BIRD

The [BIRD benchmark](https://bird-bench.github.io/) is currently the only text-to-SQL benchmark that is supported. Evaluation is based on the SQL evaluation accuracy calculated in [this file](https://github.com/AlibabaResearch/DAMO-ConvAI/blob/main/bird/llm/src/evaluation.py) provided in the BIRD GitHub repository.

#### Data Preparation

First, the data must be downloaded and prepared, which you can do by running:
```bash
ns prepare_data birdbench
```

This will download and unpack a file into `<output_directory>/dev_20240627`, which contains the BIRD dev manifest, table information, and database schemas. By default, `output_directory` will be under `nemo_skills/dataset/birdbench/`, though this can be changed via command line argument.

> The script will also process the original manifest into `<output_directory>/dev.jsonl`, which will be the input for evaluation.

Author: Removed!

#### Running the Evaluation

The following command runs an evaluation of [Qwen3-8B](https://huggingface.co/Qwen/Qwen3-8B) on a Slurm cluster.

```bash
ns eval \
     --cluster=<CLUSTER_NAME> \
     --server_type='sglang' \
     --server_gpus=8 \
     --model=Qwen/Qwen3-8B \
     --benchmarks=birdbench \
     --data_dir=<DATA_DIR> \
     --output_dir=<OUTPUT_DIR> \
     ++inference.tokens_to_generate=10000 \
     ++inference.temperature=0.6 \
     ++inference.top_p=0.95 \
     ++inference.top_k=20 \
     ++max_concurrent_requests=1024 \
```

> You should specify: `<CLUSTER_NAME>`, which should match your cluster config name; `<DATA_DIR>`, which should be the path to your base `nemo_skills/datasets` directory on the cluster; and `<OUTPUT_DIR>`.

Reviewer: ```suggestion
You should specify: `<CLUSTER_NAME>`, which should match your cluster config name; `<DATA_DIR>`, which should be a mounted path on the cluster where that datasets can be saved; and `<OUTPUT_DIR>`.
```

Reviewer: can you also maybe link to here? https://nvidia-nemo.github.io/Skills/evaluation/#using-data-on-cluster

#### Data Preparation

First, the data must be downloaded and prepared, which you can do by running:
```bash
ns prepare_data birdbench
```

This will download and unpack a file into `<output_directory>/dev_20240627`, which contains the BIRD dev manifest, table information, and database schemas. By default, `output_directory` will be under `nemo_skills/dataset/birdbench/`, though this can be changed via command line argument.

The script will also process the original manifest into `<output_directory>/dev.jsonl`, which will be the input for evaluation.

#### Running the Evaluation

The following command runs an evaluation of [Qwen3-8B](https://huggingface.co/Qwen/Qwen3-8B) on a Slurm cluster.

```bash
ns eval \
     --cluster=<CLUSTER_NAME> \
     --server_type='sglang' \
     --server_gpus=8 \
     --model=Qwen/Qwen3-8B \
     --benchmarks=birdbench \
     --output_dir=<OUTPUT_DIR> \

>      --installation_command="pip install func_timeout" \

Author: Oh good catch, thanks. Removing these references to `installation_command`.

     ++inference.tokens_to_generate=10000 \
     ++inference.temperature=0.6 \
     ++inference.top_p=0.95 \
     ++inference.top_k=20 \
     ++max_concurrent_requests=1024 \
     ++eval_config.dev_json_filepath=<DATA_DIR>/dev.json \
     ++eval_config.db_path=<DATA_DIR>/dev_databases
```
!!! note

>     BIRD evaluation requires the `func_timeout` library, so you will need to add the installation argument `--installation_command` seen in this example if it is not included in your container.

Reviewer: again probably unnecessary once we merge this

> The arguments for evaluation also include paths for `++eval_config.dev_json_filepath` and `++eval_config.db_path`. These are used for SQL execution. You should point these to the unmodified `dev.json` file and `dev_databases` directory found in the original downloaded and unzipped data (i.e. `.../dev_20240627/dev.json`, etc.). If using a Slurm cluster, you may need to upload and mount the data directory first.

Author: (Side note: I removed `dev_json_filepath` in my last refactor so at least that's cleaner now :)

> The arguments for evaluation also include paths for `++eval_config.dev_json_filepath` and `++eval_config.db_path`. These are used for SQL execution. You should point these to the unmodified `dev.json` file and `dev_databases` directory found in the original downloaded and unzipped data (i.e. `.../dev_20240627/dev.json`, etc.). If using a Slurm cluster, you may need to upload and mount the data directory first.

Reviewer: Can we use the `data_dir` argument instead, then?

https://github.com/NVIDIA-NeMo/Skills/blob/main/nemo_skills/pipeline/prepare_data.py#L34-L50

Then we just have to pass around that same argument between prepare_data and eval, but don't need to go separately prepare it on the cluster and then get the path back. Makes it easier to keep things consistent.

> The arguments for evaluation also include paths for `++eval_config.dev_json_filepath` and `++eval_config.db_path`. These are used for SQL execution. You should point these to the unmodified `dev.json` file and `dev_databases` directory found in the original downloaded and unzipped data (i.e. `.../dev_20240627/dev.json`, etc.). If using a Slurm cluster, you may need to upload and mount the data directory first.

Author: I think I might be missing something here but I'm still not quite seeing how that would work--if we ran prepare_data locally and then wanted to run evaluation on the cluster, at prepare_data runtime `data_dir` would only have the local path to the directory and not the cluster's path, right? Additionally, if we don't want to have NeMo-Skills upload the dataset for every run, we can't guarantee it'll be in any given relative path location e.g. `nemo_skills/datasets/birdbench/...` since the user could upload it anywhere else.

(And a quick clarification: currently we don't need to prepare the data separately, the pipeline is more like (1) run prepare_data locally, (2) upload that directory to desired cluster, (3) run evaluation with that directory mounted & pass in that path.)
